### PR TITLE
[PLAT-424] active season

### PR DESCRIPTION
**Pull request by raulmartinezm**
## Description

Please describe your pull request and link relevant issues.

## Type of changes

- [ ] `build`: Changes that affect the build system or external dependencies (eg, Cargo, Docker)
- [ ] `ci`: Changes to CI configuration
- [ ] `docs`: Changes to documentation only
- [x] `feat`: Changes to add a new feature
- [ ] `fix`: Changes to fix a bug
- [ ] `refactor`: Changes that do not alter functionality
- [ ] `style`: Changes to format the code
- [ ] `test`: Changes to add missing tests or correct existing tests

## Checklist

- [x] Tests for the changes have been added
- [ ] Necessary documentation is added (if appropriate)
- [x] Formatted with `cargo fmt --all`
- [x] Linted with `cargo clippy --all-features --all-targets`
- [x] Tested with `cargo test --workspace --all-features --all-targets`

### CODE REVIEW

File: pallets/ajuna-awesome-avatars/src/lib.rs [hunk 1]
```diff
@@ -271,4 +287,32 @@ pub mod pallet {
...
 		}
 	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(block_number: T::BlockNumber) -> Weight {
+			let active_season_id = Self::active_season_id();
+			let next_season_id = Self::next_active_season_id();
+			let season_id = active_season_id.unwrap_or(next_season_id);
+			let maybe_season = Self::seasons(season_id);
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/lib.rs:297]: Let's remove some variables and add `db_weight` here:
```suggestion
			let season_id = Self::active_season_id().unwrap_or(Self::next_active_season_id());
			let mut db_weight = T::DbWeight::get().reads(2);
```

File: pallets/ajuna-awesome-avatars/src/lib.rs [hunk 2]
```diff
@@ -271,4 +287,32 @@ pub mod pallet {
...
+			if let Some(season) = maybe_season {
+				if season.early_start <= block_number && block_number <= season.end {
+					ActiveSeasonId::<T>::put(season_id);
+					if block_number >= season.end {
+						NextActiveSeasonId::<T>::mutate(|season_id| season_id.saturating_inc());
+					}
+				} else {
+					ActiveSeasonId::<T>::kill();
+				}
+			}
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/lib.rs:308]: I think I made a mistake in the `NextActiveSeasonId` logic. We just need to put `season_id + 1` to just serve it as a lookup table when `ActiveSeasonId` is `None`:
```suggestion
			if let Some(season) = Self::seasons(season_id) {
				db_weight += T::DbWeight::get().reads(1);
				if season.early_start <= now && now <= season.end {
					ActiveSeasonId::<T>::put(season_id);
					NextActiveSeasonId::<T>::put(season_id.saturating_add(1));
					db_weight += T::DbWeight::get().writes(2);
				} else {
					ActiveSeasonId::<T>::kill();
					db_weight += T::DbWeight::get().writes(1);
				}
			}
			db_weight
```

File: pallets/ajuna-awesome-avatars/src/mock.rs [hunk 3]
```diff
@@ -15,7 +15,8 @@
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 use crate::{self as pallet_ajuna_awesome_avatars, season::Season};
-use frame_support::traits::{ConstU16, ConstU64};
+use frame_support::traits::{ConstU16, ConstU64, Hooks};
+
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/mock.rs:19]:
```suggestion
```

File: pallets/ajuna-awesome-avatars/src/tests.rs [hunk 4]
```diff
@@ -615,6 +615,61 @@ mod season {
 			);
 		});
 	}
+
+	#[test]
+	fn active_season_hooks_should_update_active_season_on_block_initialize() {
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/tests.rs:620]: For happy paths, let's stick to `should_work` for consistency
```suggestion
	fn active_season_hooks_should_work() {
```

File: pallets/ajuna-awesome-avatars/src/tests.rs [hunk 5]
```diff
@@ -615,6 +615,61 @@ mod season {
...
+					max_mints: 1,
+					max_mythical_mints: 1,
+					rarity_tiers: get_rarity_tiers(),
+					max_variations: 1
+				}
+			));
+
+			run_to_block(2);
+			assert!(AwesomeAvatars::active_season_id().is_some());
+			assert_eq!(AwesomeAvatars::active_season_id().unwrap(), 1);
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/tests.rs:651]: Let's simplify these lines into:
```suggestion
			assert_eq!(AwesomeAvatars::active_season_id(), Some(1));
```

File: pallets/ajuna-awesome-avatars/src/tests.rs [hunk 6]
```diff
@@ -615,6 +615,61 @@ mod season {
...
+
+			run_to_block(15);
+			assert!(AwesomeAvatars::active_season_id().is_some());
+			assert_eq!(AwesomeAvatars::active_season_id().unwrap(), 2);
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/tests.rs:655]:
```suggestion
			assert_eq!(AwesomeAvatars::active_season_id(), Some(2));
```

File: pallets/ajuna-awesome-avatars/src/tests.rs [hunk 7]
```diff
@@ -615,6 +615,61 @@ mod season {
 			);
 		});
 	}
+
+	#[test]
+	fn active_season_hooks_should_update_active_season_on_block_initialize() {
+		ExtBuilder::default().organizer(ALICE).build().execute_with(|| {
+			assert!(AwesomeAvatars::active_season_id().is_none());
+
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/tests.rs:623]: We also need to check whether:
- `NextActiveSeasonId` is / isn't updated correctly
- `ActiveSeasonId` is `None` if out of season

File: pallets/ajuna-awesome-avatars/src/tests.rs [hunk 8]
```diff
@@ -615,6 +615,61 @@ mod season {
...
+			assert!(AwesomeAvatars::active_season_id().is_some());
+			assert_eq!(AwesomeAvatars::active_season_id().unwrap(), 2);
+		});
+	}
+
+	#[test]
+	fn active_season_hooks_should_do_nothing_if_no_season_created() {
+		ExtBuilder::default().organizer(ALICE).build().execute_with(|| {
+			assert!(AwesomeAvatars::active_season_id().is_none());
+			assert_eq!(AwesomeAvatars::next_active_season_id(), 1);
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/tests.rs:663]: For consistency:
```suggestion
			assert_eq!(AwesomeAvatars::next_active_season_id(), 1);

```

File: pallets/ajuna-awesome-avatars/src/tests.rs [hunk 9]
```diff
@@ -615,6 +615,61 @@ mod season {
...
+			assert_eq!(AwesomeAvatars::active_season_id().unwrap(), 1);
+
+			run_to_block(15);
+			assert!(AwesomeAvatars::active_season_id().is_some());
+			assert_eq!(AwesomeAvatars::active_season_id().unwrap(), 2);
+		});
+	}
+
+	#[test]
+	fn active_season_hooks_should_do_nothing_if_no_season_created() {
```
Comment (cowboy-bebug) [inline pallets/ajuna-awesome-avatars/src/tests.rs:660]:
```suggestion
	fn active_season_hooks_should_do_nothing_if_no_season_exists() {
```

**Review by cowboy-bebug — CHANGES_REQUESTED** [review]
Looks good so far. Requesting some changes

File: pallets/ajuna-awesome-avatars/src/tests.rs (hunk 9 above)
Comment (raulmartinezm) [reply pallets/ajuna-awesome-avatars/src/tests.rs:660]: Done

File: pallets/ajuna-awesome-avatars/src/tests.rs (hunk 6 above)
Comment (raulmartinezm) [reply pallets/ajuna-awesome-avatars/src/tests.rs:655]: Changed

File: pallets/ajuna-awesome-avatars/src/tests.rs (hunk 5 above)
Comment (raulmartinezm) [reply pallets/ajuna-awesome-avatars/src/tests.rs:651]: Changed!

File: pallets/ajuna-awesome-avatars/src/lib.rs (hunk 2 above)
Comment (raulmartinezm) [reply pallets/ajuna-awesome-avatars/src/lib.rs:308]: Much better this way, changed!

File: pallets/ajuna-awesome-avatars/src/lib.rs (hunk 1 above)
Comment (raulmartinezm) [reply pallets/ajuna-awesome-avatars/src/lib.rs:297]: done

File: pallets/ajuna-awesome-avatars/src/lib.rs [hunk 10]
```diff
@@ -271,4 +287,32 @@ pub mod pallet {
...
+
+			// Register the Weight used on_finalize.
+			// 	- One storage read to get the block_weight.
+			// 	- One storage read to get the Elasticity.
+			// 	- One write to BaseFeePerGas.
+			let db_weight = <T as frame_system::Config>::DbWeight::get();
+			db_weight.reads(2).saturating_add(db_weight.write)
```
[the remark this answers is not in the document]
Comment (raulmartinezm) [reply pallets/ajuna-awesome-avatars/src/lib.rs:315]: Done

File: pallets/ajuna-awesome-avatars/src/lib.rs [hunk 11]
```diff
@@ -271,4 +287,32 @@ pub mod pallet {
 			Ok(())
 		}
 	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(block_number: T::BlockNumber) -> Weight {
```
[the remark this answers is not in the document]
Comment (raulmartinezm) [reply pallets/ajuna-awesome-avatars/src/lib.rs:293]: Changed!

**Review by cowboy-bebug — APPROVED** [review]
Nice work :shipit:

**Pull request merged**